src/HexagonOptimize.cpp

```diff
@@ -0,0 +1,1084 @@
+}
+
+namespace {
+
+// This mutator rewrites patterns with an unknown number of lanes to
+// have the specified number of lanes.
+class WithLanes : public IRMutator {
+    using IRMutator::visit;
+
+    int lanes;
+
+    Type with_lanes(Type t) { return t.with_lanes(lanes); }
+
+    void visit(const Cast *op) {
+        if (op->type.lanes() != lanes) {
+            expr = Cast::make(with_lanes(op->type), mutate(op->value));
+        } else {
+            IRMutator::visit(op);
+        }
+    }
+
+    void visit(const Variable *op) {
+        if (op->type.lanes() != lanes) {
+            expr = Variable::make(with_lanes(op->type), op->name);
+        } else {
+            expr = op;
+        }
+    }
+
+    void visit(const Broadcast *op) {
+        if (op->type.lanes() != lanes) {
+            expr = Broadcast::make(op->value, lanes);
+        } else {
+            IRMutator::visit(op);
+        }
+    }
+
+public:
+    WithLanes(int lanes) : lanes(lanes) {}
+};
+
+Expr with_lanes(Expr x, int lanes) {
+    return WithLanes(lanes).mutate(x);
+}
+
+Expr u8(Expr E) { return cast(UInt(8, E.type().lanes()), E); }
+Expr i8(Expr E) { return cast(Int(8, E.type().lanes()), E); }
+Expr u16(Expr E) { return cast(UInt(16, E.type().lanes()), E); }
+Expr i16(Expr E) { return cast(Int(16, E.type().lanes()), E); }
+Expr u32(Expr E) { return cast(UInt(32, E.type().lanes()), E); }
+Expr i32(Expr E) { return cast(Int(32, E.type().lanes()), E); }
+Expr u64(Expr E) { return cast(UInt(64, E.type().lanes()), E); }
+Expr i64(Expr E) { return cast(Int(64, E.type().lanes()), E); }
+Expr bc(Expr E) { return Broadcast::make(E, 0); }
+
+Expr min_i8 = i8(Int(8).min());
+Expr max_i8 = i8(Int(8).max());
+Expr min_u8 = u8(UInt(8).min());
+Expr max_u8 = u8(UInt(8).max());
+Expr min_i16 = i16(Int(16).min());
+Expr max_i16 = i16(Int(16).max());
+Expr min_u16 = u16(UInt(16).min());
+Expr max_u16 = u16(UInt(16).max());
+Expr min_i32 = i32(Int(32).min());
+Expr max_i32 = i32(Int(32).max());
+Expr min_u32 = u32(UInt(32).min());
+Expr max_u32 = u32(UInt(32).max());
+
+// The simplifier eliminates max(x, 0) for unsigned x, so make sure
+// our patterns reflect the same.
+Expr simplified_clamp(Expr x, Expr min, Expr max) {
+    if (x.type().is_uint() && is_zero(min)) {
+        return Halide::min(x, max);
+    } else {
+        return clamp(x, min, max);
+    }
+}
+
+Expr i32c(Expr e) { return i32(simplified_clamp(e, min_i32, max_i32)); }
+Expr u32c(Expr e) { return u32(simplified_clamp(e, min_u32, max_u32)); }
+Expr i16c(Expr e) { return i16(simplified_clamp(e, min_i16, max_i16)); }
+Expr u16c(Expr e) { return u16(simplified_clamp(e, min_u16, max_u16)); }
+Expr i8c(Expr e) { return i8(simplified_clamp(e, min_i8, max_i8)); }
+Expr u8c(Expr e) { return u8(simplified_clamp(e, min_u8, max_u8)); }
+
+struct Pattern {
+    enum Flags {
+        InterleaveResult = 1 << 0,  ///< After evaluating the pattern, interleave native vectors of the result.
```

Comment:
I think these should not be doxygen comments, since this is in an implementation file.